src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -118,9 +126,16 @@ NativeImage *NativeImage::Open(
     {
         COMPlusThrowHR(COR_E_BADIMAGEFORMAT);
     }
-    NewHolder<NativeImage> image = new NativeImage(peLoadedImage.Extract(), nativeImageFileName);
-    image->Initialize(pHeader, pLoaderAllocator, pamTracker);
-    return image.Extract();
+    NewHolder<NativeImage> image = new NativeImage(pAssemblyLoadContext, peLoadedImage.Extract(), nativeImageFileName);
+    image->Initialize(pHeader, pLoaderAllocator);
+    pExistingImage = AppDomain::GetCurrentDomain()->SetNativeImage(peLoadedImage->GetBase(), image);
```

Comment:
The regular R2R images use `READYTORUN_HELPER_Module` fixup to claim the image. Look for `AcquireImage` in readytoruninfo.cpp. Would be make sense to use the same scheme here instead of maintaining a hashtable?

Reply:
Hmm, I'm afraid that, according to your other comment, we still need some table on Linux to cater for the fact that repeated dll loads allocate separate instances. In combination with David's suggestion it seems to me that it would be easiest to keep the dictionary in AppDomain (to let us detect and fail fast upon an attempt to load the composite DLL into a secondary ALC on Linux), just key it off the composite image simple name and apply the path search when opening the dll for the first time.

Reply:
Resolved in 2nd commit as described in my other comment.